### ISSUE TITLE
Proposal: support for building artefacts based on proxy.golang.org-resolved modules

### ISSUE THREAD

**Issue by myitcv** [goreleaser/goreleaser#1354]
```
goreleaser version: v0.126.0
$ go version
go version devel +1cd724acb6 Tue Feb 18 20:21:35 2020 +0000 linux/amd64
```

**Is your feature request related to a problem? Please describe.**

I followed the quickstart guide on https://goreleaser.com/ and created a v0.1.0 release of a test program:

https://github.com/myitcvscratch/goreleasertest/releases/tag/v0.1.0

The resulting binary (for Linux AMD64) does not, however, have module information baked in:

```
cd $(mktemp -d)
curl -sL https://github.com/myitcvscratch/goreleasertest/releases/download/v0.1.0/goreleasertest_0.1.0_Linux_x86_64.tar.gz | tar -zx
go version -m ./goreleasertest
```

gives:

```
./goreleasertest: devel +1cd724acb6 Tue Feb 18 20:21:35 2020 +0000
        path    github.com/myitcvscratch/goreleasertest
        mod     github.com/myitcvscratch/goreleasertest (devel)
```

**Describe the solution you'd like**

To be able to build release artefacts from modules pulled via proxy.golang.org such that module information is baked in to the resulting binaries. 

e.g. something like:

```
$ goreleaser --proxy github.com/myitcvscratch/goreleasertest@v0.1.0
```

would result in GitHub artefacts being created based on proxy-resolved module which module information baked into the resulting binaries such that:

```
$ go version -m ./goreleasertest
/tmp/tmp.TBNzsd1QJx/bin/goreleasertest: devel +1cd724acb6 Tue Feb 18 20:21:35 2020 +0000
        path    github.com/myitcvscratch/goreleasertest
        mod     github.com/myitcvscratch/goreleasertest v0.1.0  h1:LNFqkSDf5s1eVcYh5OW2mfNWugS6oMvraH7C5TbPw5A=
```

To my mind this resolves an important skew that can be introduced between the module proxy and a GitHub tag (although the GitHub tag can still, subsequently, be moved/deleted). But more importantly bakes in exact module and authentication information.

Having the module version details be baked into the resulting binary is also consistent with approaches taken by `gopls` and other projects.

**Describe alternatives you've considered**

Doing things by hand, but that really defeats the point of using `goreleaser`!

**Additional context**

n/a

**Comment by caarlos0** [comment]
This would require way too many changes on goreleaser.

["Yes is forever"](https://twitter.com/solomonstre/status/715277134978113536), so, I rather not do this - or at least not right now.

Thanks for the suggestion either way.

Cheers!

**Comment by caarlos0** [comment]
https://twitter.com/_myitcv/status/1367486387289751552?s=21 

@myitcv pinged me on twitter about this

**Comment by caarlos0** [comment]
What do you think the impl should look like?

**Comment by myitcv** [comment]
Thanks for reconsidering, @caarlos0.

> What do you think the impl should look like?

Some variation on the following should work. Consider the example of `goreleaser` being asked to create release artefacts for the CUE project at `v0.3.0-beta.5`. The single artefact that needs to be built in that case is `cuelang.org/go/cmd/cue` so I will just consider that for now.

```
export GOPROXY="https://proxy.golang.org,direct"
export GOSUMDB="sum.golang.org"
export GONOPROXY=""
export GONOSUMDB=""
export GOPRIVATE=""
export GOINSECURE=""
cd $(mktemp -d)
go mod init blah.com
cat <<'EOD' > main.go
// +build main

package main

import _ "cuelang.org/go/cmd/cue"
EOD
go get -d cuelang.org/go/cmd/cue@v0.3.0-beta.5
go mod tidy
go build -o cue cuelang.org/go/cmd/cue
```

Quickly examining the results:

```
go version -m cue
```

gives:

```
cue: go1.16
        path    cuelang.org/go/cmd/cue
        mod     cuelang.org/go  v0.3.0-beta.5   h1:c+zS9MBCFzbuz+RI+3dp2PwfnS05aF3MqSEaAGFwzSk=
        dep     github.com/cockroachdb/apd/v2   v2.0.1  h1:y1Rh3tEU89D+7Tgbw+lp52T6p/GJLpDmNvr10UWqLTE=
        dep     github.com/emicklei/proto       v1.6.15 h1:XbpwxmuOPrdES97FrSfpyy67SSCV/wBIKXqgJzh6hNw=
        dep     github.com/mpvl/unique  v0.0.0-20150818121801-cbe035fff7de      h1:D5x39vF5KCwKQaw+OC9ZPiLVHXz3UFw2+psEX+gYcto=
        dep     github.com/pkg/errors   v0.8.1  h1:iURUrRGxPUNPdy5/HRSm+Yj6okJ6UtLINN0Q9M4+h3I=
        dep     github.com/spf13/cobra  v1.0.0  h1:6m/oheQuQ13N9ks4hubMG6BnvwOeaJrqSPLahSnczz8=
        dep     github.com/spf13/pflag  v1.0.3  h1:zPAT6CGy6wXeQ7NtTnaTerfKOsV6V6F8agHXFiazDkg=
        dep     golang.org/x/mod        v0.3.1-0.20200828183125-ce943fd02449    h1:xUIPaMhvROX9dhPvRCenIJtU78+lbEenGbgqB5hfHCQ=
        dep     golang.org/x/net        v0.0.0-20200226121028-0de0cce0169b      h1:0mm1VjtFUOIlE1SbDlwjYaDxZVDP2S5ou6y0gSgXHu8=
        dep     golang.org/x/text       v0.3.2  h1:tW2bmiBqwgJj/UpqtC8EpXEZVYOwU0yG4iWbprSVAcs=
        dep     golang.org/x/tools      v0.0.0-20200612220849-54c614fe050c      h1:g6oFfz6Cmw68izP3xsdud3Oxu145IPkeFzyRg58AKHM=
        dep     golang.org/x/xerrors    v0.0.0-20191204190536-9bdfabe68543      h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
        dep     gopkg.in/yaml.v3        v3.0.0-20200121175148-a6ecf24a6d71      h1:Xe2gvTZUJpsvOWUnvmL/tmhVBZUmHSvLbMjRj6NUUKo=
```

as required/expected. Note this output is basically what makes a [build verifiable](https://research.swtch.com/vgo-repro).

Some notes on the above:

* I might well skip over important details because I'm not 100% familiar with `goreleaser`, so apologies for that!
* I'm using a temporary directory in order to share reproducible steps; that's not critical to the process
* The temporary Go module exists purely as a way of resolving `cuelang.org/go` as a _dependency_, rather than it being the main module which would be the case if you clone a copy of the repo and built artefacts in the context of that clone
* The use of a build-ignored `main` package is not strictly necessary but probably good practice in this case. It's a way of declaring a dependency on the `cuelang.org/go/cmd/cue` main package. Clearly you can't have a `main` package dependency, hence the build constraint. But `go mod tidy` (also not strictly necessary) understands the dependency relationship and uses that as part of MVS. Having these steps in there would mean that a human could easily inspect/debug the context in which the build happened and the state would be meaningful and reproducible
* Note that if you needed to build multiple artefacts _from the same module_, they could each be added to the same `main.go` file as imports and built in the same way. But if you are building artefacts from _different_ modules, then you must use separate temporary Go modules
* The use of an explicit environment for `GOPROXY` et al ensures that we really do resolve everything through `proxy.golang.org` and `sum.golang.org`. If a user has a custom proxy/sumdb for non-public modules, that should be used in some configuration here. Private modules with no custom proxy/sumdb would likely need some setup like:

```
export GOPROXY="https://proxy.golang.org,direct"
export GOSUMDB="sum.golang.org"
export GONOPROXY=""
export GONOSUMDB=""
export GOPRIVATE="example.com/blah"
export GOINSECURE=""
```

for any `example.com/blah` being a private module. Again, very precise and prescriptive allowing of those modules that are private and hence should not be resolved via `proxy.golang.org` or `sum.golang.org`

**Comment by caarlos0** [comment]
Yeah, that definitely works. Will work on something to see how it looks like

Thanks for the throughout notes/info/etc 🤘

**Comment by caarlos0** [comment]
@myitcv if you want to take a look, very wip, would like to know your thoughts though: https://github.com/goreleaser/goreleaser/pull/2129

**Comment by myitcv** [comment]
Just noting a further suggestion here as part of some notes I'm leaving on #2129: namely that the `go.sum` file that belongs to the module containing the main package being built should be copied to the temporary module. I've updated the example in https://github.com/goreleaser/goreleaser/issues/1354#issuecomment-791175542

**Comment by caarlos0** [comment]
will be released on v0.162.0
